File: linear-sort-optimized.rs
// sort only positive numbers
// linear sort optimized for performance by halving sleep time

use std::thread;
use std::time::Duration;
use std::sync::mpsc;
use std::convert::TryInto;

fn main() {

    let array: [i32; 10] = [6, 4, 9, 2, 8, 1, 7, 5, 3 , 10];

    linear_sort(array);

}

fn linear_sort(array: [i32; 10]) {

    let (tx, rx) = mpsc::channel();

    println!("hello");
    println!("{}", array[0]);

    for val in array {
        let tx1 = tx.clone();
        thread::spawn(move || {
                thread::sleep(Duration::from_millis(500 *  <i32 as TryInto<u64>>::try_into(val).unwrap()));
                tx1.send(val).unwrap();
        });
    }

    let _ = tx.send(0);

    let mut vec = Vec::new();

    for received in rx {
        //val = rx.recv().unwrap();
        println!("Got: {received}");
        vec.push(received);
    }

    println!("{:?}", vec);

}

File: linear-sort-optimized.go
package main

import (
    "fmt"
    "time"
)

func linearSortGoRoutine(num int, c chan<- int) {
    time.Sleep(500 * time.Duration(num) * time.Millisecond)
    c <- num
}

func main() {

    array:= [...]int{8, 6, 10, 9, 5, 2, 7, 1, 4, 3}

    fmt.Println("array:", array)

    messages := make(chan int)

    for i := 0; i < len(array); i++ {
        go linearSortGoRoutine(array[i], messages)
    }

    res := []int{}

    for i:= 0; i < len(array); i++ {
        num := <-messages
        fmt.Println("Got:", num)
        res = append(res, num)
    }

    fmt.Println(res)

    fmt.Println("hello world")
}